reconstruct_surface_batches.py

```diff
@@ -0,0 +1,607 @@
+        start_idx = b * batch_size
+        end_idx = min((b + 1) * batch_size, num_patches)
+        for i in range(start_idx, end_idx):
+            dev_i = args.devices[i % len(args.devices)]
+            phi[i] = phi[i].to(dev_i)
+            patch_uvs[i] = patch_uvs[i].to(dev_i)
+            patch_xs[i] = patch_xs[i].to(dev_i)
+            pi[i] = pi[i].to(dev_i)
+        optimizer = phi_optimizers[b]
+        move_optimizer_to_device(optimizer, phi_optimizers_devices[b])
+        for epoch in range(args.global_epochs):
+            optimizer.zero_grad()
+            uv_optimizer.zero_grad()
+
+            sum_loss = torch.tensor([0.0]).to(args.devices[0])
+            epoch_start_time = time.time()
+            for i in range(start_idx, end_idx):
+                uv_i = patch_uvs[i]
+                x_i = patch_xs[i]
+                y_i = phi[i](uv_i)
+                pi_i = pi[i]
+                loss_i = mse_loss(x_i[pi_i].unsqueeze(0), y_i.unsqueeze(0))
+
+                if loss_i.item() < best_losses[i]:
+                    best_losses[i] = loss_i.item()
+                    model_copy = copy.deepcopy(phi[i]).to('cpu')
+                    best_models[i] = copy.deepcopy(model_copy.state_dict())
+
+                sum_loss += loss_i.to(args.devices[0])
+
+            sum_loss.backward()
+            epoch_end_time = time.time()
+
+            print("%d/%d: [Total = %0.5f] [Mean = %0.5f] [Time = %0.3f]" %
+                  (epoch, args.global_epochs, sum_loss.item(),
+                   sum_loss.item() / (end_idx - start_idx), epoch_end_time-epoch_start_time))
+            optimizer.step()
+            uv_optimizer.step()
+        for i in range(start_idx, end_idx):
+            dev_i = 'cpu'
+            phi[i] = phi[i].to(dev_i)
+            patch_uvs[i] = patch_uvs[i].to(dev_i)
+            patch_xs[i] = patch_xs[i].to(dev_i)
+            pi[i] = pi[i].to(dev_i)
+        move_optimizer_to_device(optimizer, 'cpu')
+                    
+    print("Mean best losses:", np.mean(best_losses[i]))
+    for i, phi_i in enumerate(phi):
+        phi_i.load_state_dict(best_models[i])
+
+    output_dict["final_model"] = phi.state_dict()
+
+    print("Generating dense point cloud...")
+    v, n = upsample_surface(patch_uvs, patch_tx, phi, args.devices,
+                            scale=(1.0/args.padding),
+                            num_samples=args.upsamples_per_patch,
+                            normal_samples=args.normal_neighborhood_size,
+                            num_batches=num_batches,
+                            compute_normals=False)
+
+    print("Saving dense point cloud...")
+    pcu.write_ply(args.output + ".ply", v, np.zeros([], dtype=np.int32), n, np.zeros([], dtype=v.dtype))
+
+    print("Saving metadata...")
+    torch.save(output_dict, args.output + ".pt")
+
+    if args.plot:
+        plot_reconstruction(x, patch_uvs, patch_tx, phi, scale=1.0/args.padding)
+
+
+if __name__ == "__main__":
+    main()
```

Comment:
Before merging, I want to make sure of a couple of things: 
* First if we are not running in batched mode, I want to check how much performance impact we incur from moving things back and forth to/from the CPU/GPU. If it doesn't take much longer to reconstruct with this script and the original script, then we can replace the old `reconstruct_surface.py` with this one. 
* The second thing I want to make sure of is that this produces the same results as the old script for the same seed. We checked this offline with an earlier version so it should be fine, but I want to verify with this version. 

If these checks are both okay, then let's replace the old `reconstruct_surface.py` with this file. :)

Reply:
Okay I ran some tests and this is looking good! It looks like for a single batch there is a small overhead over the old code but it's not a huge deal and I'd rather have a single piece of code I'm going to merge this and push some small changes on top of it.